rafx-api/src/backends/gles3/command_buffer.rs

```diff
@@ -0,0 +1,1168 @@
+    RafxExtents3D, RafxIndexBufferBinding, RafxIndexType, RafxLoadOp, RafxResourceType, RafxResult,
+    RafxTextureBarrier, RafxVertexBufferBinding, MAX_DESCRIPTOR_SET_LAYOUTS,
+};
+
+use rafx_base::trust_cell::TrustCell;
+
+use crate::gles3::conversions::{array_layer_to_cube_map_target, Gles3DepthStencilState};
+use crate::gles3::gles3_bindings;
+
+use crate::backends::gles3::{RafxRawImageGles3, RafxSamplerIndexGles3};
+use crate::gles3::gl_type_util;
+use crate::gles3::gles3_bindings::types::GLenum;
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub struct RafxCommandBufferGles3 {
+    queue: RafxQueueGles3,
+    command_pool_state: CommandPoolGles3State,
+}
+
+impl RafxCommandBufferGles3 {
+    pub(crate) fn queue(&self) -> &RafxQueueGles3 {
+        &self.queue
+    }
+
+    pub fn new(
+        command_pool: &RafxCommandPoolGles3,
+        _command_buffer_def: &RafxCommandBufferDef,
+    ) -> RafxResult<RafxCommandBufferGles3> {
+        Ok(RafxCommandBufferGles3 {
+            queue: command_pool.queue().clone(),
+            command_pool_state: command_pool.command_pool_state().clone(),
+        })
+    }
+
+    pub fn begin(&self) -> RafxResult<()> {
+        let mut state = self.command_pool_state.borrow_mut();
+        assert!(!state.is_started);
+        state.is_started = true;
+
+        Ok(())
+    }
+
+    pub fn end(&self) -> RafxResult<()> {
+        let mut state = self.command_pool_state.borrow_mut();
+        assert!(state.is_started);
+
+        let gl_context = self.queue.device_context().gl_context();
+
+        // We purposely do not clear framebuffer_color_bound, framebuffer_depth_bound, or
+        // framebuffer_stencil_bound. The framebuffer and tracking if a texture is bound should
+        // be persisted across frames. The state is private to the command pool.
+
+        state.is_started = false;
+        assert!(state.surface_size.is_none());
+        state.current_gl_pipeline_info = None;
+        state.stencil_reference_value = 0;
+        state.clear_bindings();
+        Self::update_vertex_attributes_in_use(gl_context, &mut *state, 0)?;
+        assert_eq!(state.vertex_attribute_enabled_bits, 0);
+        for attribute in &mut state.vertex_attributes {
+            *attribute = None;
+        }
+        for vertex_offset in &mut state.currently_bound_vertex_offset {
+            *vertex_offset = None;
+        }
+        for bound_vertex_buffer in &mut state.bound_vertex_buffers {
+            *bound_vertex_buffer = None;
+        }
+        state.index_buffer_byte_offset = 0;
+
+        Ok(())
+    }
+
+    pub fn return_to_pool(&self) -> RafxResult<()> {
+        // don't need to do anything
+        Ok(())
+    }
+
+    fn bind_framebuffer(
+        gl_context: &GlContext,
+        texture: &RafxTextureGles3,
+        attachment: GLenum,
+        array_slice: u16,
+        mip_slice: u8,
+    ) -> RafxResult<()> {
+        match texture.gl_raw_image() {
+            // RafxRawImageGl::Renderbuffer(id) => {
```

Comment:
We could potentially use render buffers for textures that won't be sampled. It's also possible we will need to use them to support MSAA, not sure.